File: skill/src/crypto.test.ts
import { describe, it, expect, beforeAll } from "vitest";
import { readFileSync } from "node:fs";
import { resolve } from "node:path";
import {
	ed25519PubToX25519,
	ed25519PrivToX25519,
	encrypt,
	decrypt,
	encryptWithNonce,
	ensureSodiumReady,
} from "./crypto.js";

interface IdentityVector {
	ed25519_seed: string;
	ed25519_public_key: string;
	ed25519_private_key: string;
	x25519_public_key: string;
	x25519_private_key: string;
	address: string;
}

interface CryptoVector {
	description: string;
	sender_ed25519_seed: string;
	recipient_ed25519_seed: string;
	sender_x25519_pub: string;
	sender_x25519_priv: string;
	recipient_x25519_pub: string;
	recipient_x25519_priv: string;
	nonce: string;
	plaintext: string;
	ciphertext: string;
}

function loadIdentityVectors(): IdentityVector[] {
	const data = readFileSync(
		resolve(__dirname, "../../testdata/identity_vectors.json"),
		"utf-8",
	);
	return JSON.parse(data).vectors;
}

function loadCryptoVectors(): CryptoVector[] {
	const data = readFileSync(
		resolve(__dirname, "../../testdata/crypto_vectors.json"),
		"utf-8",
	);
	return JSON.parse(data).vectors;
}

function hexToBytes(hex: string): Uint8Array {
	const bytes = new Uint8Array(hex.length / 2);
	for (let i = 0; i < hex.length; i += 2) {
		bytes[i / 2] = Number.parseInt(hex.substring(i, i + 2), 16);
	}
	return bytes;
}

function bytesToHex(bytes: Uint8Array): string {
	return Array.from(bytes)
		.map((b) => b.toString(16).padStart(2, "0"))
		.join("");
}

beforeAll(async () => {
	await ensureSodiumReady();
});

describe("Ed25519 to X25519 conversion", () => {
	it("converts public keys to match test vectors", () => {
		const vectors = loadIdentityVectors();
		for (const v of vectors) {
			const pubKey = hexToBytes(v.ed25519_public_key);
			const x25519Pub = ed25519PubToX25519(pubKey);
			expect(bytesToHex(x25519Pub)).toBe(v.x25519_public_key);
		}
	});

	it("converts private keys to match test vectors", () => {
		const vectors = loadIdentityVectors();
		for (const v of vectors) {
			const privKey = hexToBytes(v.ed25519_private_key);
			const x25519Priv = ed25519PrivToX25519(privKey);
			expect(bytesToHex(x25519Priv)).toBe(v.x25519_private_key);
		}
	});
});

describe("NaCl box encryption", () => {
	it("encrypts with known nonce to match test vectors", () => {
		const vectors = loadCryptoVectors();
		for (const v of vectors) {
			const recipPub = hexToBytes(v.recipient_x25519_pub);
			const senderPriv = hexToBytes(v.sender_x25519_priv);
			const nonce = hexToBytes(v.nonce);
			const plaintext = hexToBytes(v.plaintext);

			const sealed = encryptWithNonce(plaintext, recipPub, senderPriv, nonce);
			expect(bytesToHex(sealed)).toBe(v.ciphertext);
		}
	});

	it("decrypts known ciphertext to match test vectors", () => {
		const vectors = loadCryptoVectors();
		for (const v of vectors) {
			const senderPub = hexToBytes(v.sender_x25519_pub);
			const recipPriv = hexToBytes(v.recipient_x25519_priv);
			const ciphertext = hexToBytes(v.ciphertext);

			const decrypted = decrypt(ciphertext, senderPub, recipPriv);
			expect(bytesToHex(decrypted)).toBe(v.plaintext);
		}
	});

	it("produces different ciphertexts for same plaintext (random nonce)", async () => {
		const vectors = loadCryptoVectors();
		const v = vectors[0];
		const recipPub = hexToBytes(v.recipient_x25519_pub);
		const senderPriv = hexToBytes(v.sender_x25519_priv);
		const plaintext = hexToBytes(v.plaintext);

		const c1 = encrypt(plaintext, recipPub, senderPriv);
		const c2 = encrypt(plaintext, recipPub, senderPriv);

		expect(bytesToHex(c1)).not.toBe(bytesToHex(c2));
	});
});

File: relay/internal/crypto/crypto_test.go
package crypto_test

import (
	"crypto/ed25519"
	"encoding/hex"
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	"github.com/pinch-protocol/pinch/relay/internal/crypto"
)

type identityVector struct {
	Ed25519Seed       string `json:"ed25519_seed"`
	Ed25519PublicKey  string `json:"ed25519_public_key"`
	Ed25519PrivateKey string `json:"ed25519_private_key"`
	X25519PublicKey   string `json:"x25519_public_key"`
	X25519PrivateKey  string `json:"x25519_private_key"`
	Address           string `json:"address"`
}

type cryptoVector struct {
	Description         string `json:"description"`
	SenderSeed          string `json:"sender_ed25519_seed"`
	RecipientSeed       string `json:"recipient_ed25519_seed"`
	SenderX25519Pub     string `json:"sender_x25519_pub"`
	SenderX25519Priv    string `json:"sender_x25519_priv"`
	RecipientX25519Pub  string `json:"recipient_x25519_pub"`
	RecipientX25519Priv string `json:"recipient_x25519_priv"`
	Nonce               string `json:"nonce"`
	Plaintext           string `json:"plaintext"`
	Ciphertext          string `json:"ciphertext"`
}

func testdataDir() string {
	_, filename, _, _ := runtime.Caller(0)
	return filepath.Join(filepath.Dir(filename), "..", "..", "..", "testdata")
}

func loadIdentityVectors(t *testing.T) []identityVector {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(testdataDir(), "identity_vectors.json"))
	if err != nil {
		t.Fatalf("failed to read identity_vectors.json: %v", err)
	}
	var result struct {
		Vectors []identityVector `json:"vectors"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to parse identity_vectors.json: %v", err)
	}
	return result.Vectors
}

func loadCryptoVectors(t *testing.T) []cryptoVector {
	t.Helper()
	data, err := os.ReadFile(filepath.Join(testdataDir(), "crypto_vectors.json"))
	if err != nil {
		t.Fatalf("failed to read crypto_vectors.json: %v", err)
	}
	var result struct {
		Vectors []cryptoVector `json:"vectors"`
	}
	if err := json.Unmarshal(data, &result); err != nil {
		t.Fatalf("failed to parse crypto_vectors.json: %v", err)
	}
	return result.Vectors
}

func mustDecodeHex(t *testing.T, s string) []byte {
	t.Helper()
	b, err := hex.DecodeString(s)
	if err != nil {
		t.Fatalf("failed to decode hex %q: %v", s, err)
	}
	return b
}

func TestEd25519PublicToX25519(t *testing.T) {
	vectors := loadIdentityVectors(t)
	for _, v := range vectors {
		seed := mustDecodeHex(t, v.Ed25519Seed)
		priv := ed25519.NewKeyFromSeed(seed)
		pub := priv.Public().(ed25519.PublicKey)
		expectedX25519Pub := mustDecodeHex(t, v.X25519PublicKey)

		got, err := crypto.Ed25519PublicToX25519(pub)
		if err != nil {
			t.Fatalf("Ed25519PublicToX25519 error: %v", err)
		}
		if hex.EncodeToString(got) != hex.EncodeToString(expectedX25519Pub) {
			t.Errorf("Ed25519PublicToX25519 mismatch for seed %s:\n  got  %x\n  want %x", v.Ed25519Seed, got, expectedX25519Pub)
		}
	}
}

func TestEd25519PrivateToX25519(t *testing.T) {
	vectors := loadIdentityVectors(t)
	for _, v := range vectors {
		seed := mustDecodeHex(t, v.Ed25519Seed)
		priv := ed25519.NewKeyFromSeed(seed)
		expectedX25519Priv := mustDecodeHex(t, v.X25519PrivateKey)

		got := crypto.Ed25519PrivateToX25519(priv)
		if hex.EncodeToString(got) != hex.EncodeToString(expectedX25519Priv) {
			t.Errorf("Ed25519PrivateToX25519 mismatch for seed %s:\n  got  %x\n  want %x", v.Ed25519Seed, got, expectedX25519Priv)
		}
	}
}

func TestEncryptWithKnownNonce(t *testing.T) {
	vectors := loadCryptoVectors(t)
	for _, v := range vectors {
		t.Run(v.Description, func(t *testing.T) {
			senderPriv := mustDecodeHex(t, v.SenderX25519Priv)
			recipPub := mustDecodeHex(t, v.RecipientX25519Pub)
			nonce := mustDecodeHex(t, v.Nonce)
			plaintext := mustDecodeHex(t, v.Plaintext)
			expectedCiphertext := mustDecodeHex(t, v.Ciphertext)

			var recipPubArr, senderPrivArr [32]byte
			copy(recipPubArr[:], recipPub)
			copy(senderPrivArr[:], senderPriv)

			var nonceArr [24]byte
			copy(nonceArr[:], nonce)

			got := crypto.EncryptWithNonce(plaintext, &recipPubArr, &senderPrivArr, &nonceArr)
			if hex.EncodeToString(got) != hex.EncodeToString(expectedCiphertext) {
				t.Errorf("EncryptWithNonce mismatch:\n  got  %x\n  want %x", got, expectedCiphertext)
			}
		})
	}
}

func TestDecrypt(t *testing.T) {
	vectors := loadCryptoVectors(t)
	for _, v := range vectors {
		t.Run(v.Description, func(t *testing.T) {
			senderPub := mustDecodeHex(t, v.SenderX25519Pub)
			recipPriv := mustDecodeHex(t, v.RecipientX25519Priv)
			ciphertext := mustDecodeHex(t, v.Ciphertext)
			expectedPlaintext := mustDecodeHex(t, v.Plaintext)

			var senderPubArr, recipPrivArr [32]byte
			copy(senderPubArr[:], senderPub)
			copy(recipPrivArr[:], recipPriv)

			got, err := crypto.Decrypt(ciphertext, &senderPubArr, &recipPrivArr)
			if err != nil {
				t.Fatalf("Decrypt error: %v", err)
			}
			if hex.EncodeToString(got) != hex.EncodeToString(expectedPlaintext) {
				t.Errorf("Decrypt mismatch:\n  got  %x\n  want %x", got, expectedPlaintext)
			}
		})
	}
}

func TestEncryptRandomNonce(t *testing.T) {
	var recipPub, senderPriv [32]byte
	// Use first crypto vector's keys
	vectors := loadCryptoVectors(t)
	copy(recipPub[:], mustDecodeHex(t, vectors[0].RecipientX25519Pub))
	copy(senderPriv[:], mustDecodeHex(t, vectors[0].SenderX25519Priv))

	plaintext := []byte("same plaintext")
	c1, err := crypto.Encrypt(plaintext, &recipPub, &senderPriv)
	if err != nil {
		t.Fatalf("Encrypt error: %v", err)
	}
	c2, err := crypto.Encrypt(plaintext, &recipPub, &senderPriv)
	if err != nil {
		t.Fatalf("Encrypt error: %v", err)
	}

	if hex.EncodeToString(c1) == hex.EncodeToString(c2) {
		t.Error("encrypting same plaintext twice produced identical ciphertexts (nonce reuse!)")
	}
}

func TestNoncePrependedToCiphertext(t *testing.T) {
	vectors := loadCryptoVectors(t)
	v := vectors[0]

	senderPriv := mustDecodeHex(t, v.SenderX25519Priv)
	recipPub := mustDecodeHex(t, v.RecipientX25519Pub)
	nonce := mustDecodeHex(t, v.Nonce)
	plaintext := mustDecodeHex(t, v.Plaintext)

	var recipPubArr, senderPrivArr [32]byte
	copy(recipPubArr[:], recipPub)
	copy(senderPrivArr[:], senderPriv)

	var nonceArr [24]byte
	copy(nonceArr[:], nonce)

	sealed := crypto.EncryptWithNonce(plaintext, &recipPubArr, &senderPrivArr, &nonceArr)

	// Verify first 24 bytes are the nonce
	if len(sealed) < 24 {
		t.Fatalf("sealed too short: %d bytes", len(sealed))
	}
	if hex.EncodeToString(sealed[:24]) != hex.EncodeToString(nonce) {
		t.Errorf("first 24 bytes are not the nonce:\n  got  %x\n  want %x", sealed[:24], nonce)
	}
}
